### ISSUE TITLE
[Feature Request] Store txn sender and seq num in events

### ISSUE THREAD

**Issue by davidiw** [aptos-labs/aptos-core#745]
Right now, it is impossible to tell what triggered an event. One can add a lot of metadata to the event, but that doesn't make it systematic and makes it difficult to add super awesome APIs.

I propose we extend all events with either version -- if we can get that before too late -- or txn sender and seq number. Then we can also have REST API that returns back the timestamp of an event. Otherwise events are only marginally useful or require a lot of extra metadata.

**Comment by msmouse** [comment]
Event is bound to a version already. It's the (DB and Rest) API to be updated if we want to support this, and indices to make it more efficient.

**Comment by davidiw** [comment]
> Event is bound to a version already. It's the (DB and Rest) API to be updated if we want to support this, and indices to make it more efficient.

can you sketch out the path in the code? Or where in the DB this is? I didn't see it in ContractEvent -- or did I glance too quickly.

**Comment by sitalkedia** [comment]
@davidiw - The DB stores the events indexed by the corresponding txn version (https://github.com/aptos-labs/aptos-core/blob/main/storage/aptosdb/src/schema/event/mod.rs#L26) and seems like The API from DB already returns the event and version, we can update the rest APIs to include this as well -  https://github.com/aptos-labs/aptos-core/blob/main/storage/aptosdb/src/lib.rs#L890

**Comment by Kabie** [comment]
I really need this so I tried on my own. But now I need some help.

I tried to add a `version` field on `aptos_api_types::transaction::Event`. Most of the places looks fine, the only problem is that the `DirectWriteSet` also have an `events` field, which could be used in [transaction apis](https://fullnode.devnet.aptoslabs.com/spec.html#/schemas/PendingTransaction). Since those events can't have versions (cause they are sent by user or pending), how should I do? One thought is to use `Option<U64>` as version type in `Event`. Or use different Event types for pending and committed transaction? 

https://github.com/aptos-labs/aptos-core/blob/13cc8d563395840fe6b344425c961ea4b177a479/api/types/src/transaction.rs#L450-L453
